typing/datarepr.ml

```diff
@@ -48,6 +48,29 @@ let constructor_has_optional_shape ({cstr_attributes = attrs} : constructor_desc
   List.exists (fun (x,_) -> x.txt = internal_optional) attrs
 
 
+(**  [Types.constructor_description]
+     records the type at the definition type so for ['a option]
+     it will always be [Tvar]
+*)
+let cannot_inhabit_none_like_value (typ : Types.type_expr) =
+  match (Btype.repr typ).desc with
+  |  Tconstr(p, _,_) ->
+      if Path.same p Predef.path_option then false
+      else Predef.type_is_builtin_path p
+  | Ttuple _
+  | Tvariant _
+  | Tpackage _ 
+  | Tarrow _ -> true
+  | Tfield _ 
+  | Tpoly _ 
+  | Tunivar _ 
+  | Tlink _ 
+  | Tsubst _
+  | Tnil 
+  | Tvar _
+  | Tobject _ 
```

Comment:
Can a `Tobject` be an optional?

Reply:
I am not sure, but Tobject is rarely used, lets be conservative here